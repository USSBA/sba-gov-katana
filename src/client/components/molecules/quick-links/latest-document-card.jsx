import React from 'react'
import s from './latest-document-card.scss'

class LatestDocumentCard extends React.Component {

  render() {
     return (
       <div>
         LatestDocumentCard  component
       </div>
     );
  }
}

export default LatestDocumentCard 