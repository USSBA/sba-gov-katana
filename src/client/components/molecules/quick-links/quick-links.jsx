import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import queryString from 'querystring'
import { chain, isEmpty } from 'lodash'
import { DecorativeDash, Link } from 'atoms'
import styles from './quick-links.scss'
import { fetchSiteContent } from '../../../fetch-content-helper'

const MAX_TITLE_LENGTH = 80

function getCurrentFile(files) {
  let found = null

  if (files) {
    found = chain(files)
      .filter(item => {
        return moment(item.effectiveDate).isSameOrBefore(moment())
      })
      .sortBy('effectiveDate')
      .last()
      .value()
  }

  return found
}

function formatDate(date) {
  return moment(date).format('MMM D, YYYY')
}

const valueOrAll = value => {
  return isEmpty(value[0]) ? 'all' : value[0]
}

class QuickLinks extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      articles: [],
      documents: [],
      LOADING_STATE: 'unloaded'
    }
  }

  async componentDidMount() {
    this.setState(
      {
        LOADING_STATE: 'isLoading'
      },
      async () => {
        this.setState({
          documents: await this.fetchDocuments(),
          articles: await this.fetchArticles(),
          LOADING_STATE: 'isLoaded'
        })
      }
    )
  }

  async fetchDocuments() {
    const results = {}
    const { typeOfLinks } = this.props.data

    for (const [index, quickLink] of typeOfLinks.entries()) {
      const { documentActivity, documentOffice, documentType, documentProgram, type } = quickLink
      if (type === 'documentLookup') {
        results[`documents-${index}`] = await fetchSiteContent('documents', {
          sortBy: 'Last Updated',
          type: valueOrAll(documentType),
          program: valueOrAll(documentProgram),
          activity: valueOrAll(documentActivity),
          office: documentOffice || 'all',
          start: 0,
          end: 3
        })
      }
    }

    return results
  }

  async fetchArticles() {
    const results = {}
    const { typeOfLinks } = this.props.data

    for (const [index, quickLink] of typeOfLinks.entries()) {
      const { articleCategory, articleProgram, type } = quickLink

      if (type === 'articleLookup') {
        results[`articles-${index}`] = await fetchSiteContent('articles', {
          sortBy: 'Last Updated',
          articleCategory: valueOrAll(articleCategory),
          program: valueOrAll(articleProgram),
          start: 0,
          end: 3
        })
      }
    }

    return results
  }

  generateGrid(length) {
    const gridClass = { 1: styles.oneCard, 2: styles.twoCards }
    return length > 2 ? styles.manyCards : gridClass[length]
  }

  renderQuickLinks() {
    const {
      data: { typeOfLinks }
    } = this.props
    const { documents, articles, LOADING_STATE } = this.state
    const gridClass = this.generateGrid(typeOfLinks.length)

    return typeOfLinks.map((quickLink, index) => {
      if (quickLink.type === 'documentLookup') {
        return (
          <LatestDocumentsCard
            key={index}
            className={styles.card + ' ' + gridClass}
            documents={documents['documents-' + index]}
            LOADING_STATE={LOADING_STATE}
            {...quickLink}
          />
        )
      } else if (quickLink.type === 'ratesList') {
        return <RatesCard key={index} {...quickLink} className={styles.card + ' ' + gridClass} />
      } else if (quickLink.type === 'articleLookup') {
        return (
          <ArticlesCard
            key={index}
            className={styles.card + ' ' + gridClass}
            articles={articles['articles-' + index]}
            {...quickLink}
          />
        )
      }

      return null
    })
  }

  render() {
    return (
      <div className={styles.collection} data-testid="quick-links">
        {this.props.data ? this.renderQuickLinks() : <div>loading</div>}
      </div>
    )
  }
}

const LatestDocumentsCard = props => {
  const eventCategory = `${props.sectionHeaderText.toLowerCase()}-module`
  const { LOADING_STATE } = props
  return (
    <div data-testid="documents-card" className={props.className}>
      <div className={styles.titleContainer}>
        <h4 data-testid="header" className={styles.title}>
          {props.sectionHeaderText}
        </h4>
        <Link
          data-testid="see-all-link"
          to={`/document?${queryString.stringify({
            type: props.documentType,
            program: props.documentProgram,
            office: props.documentOffice || 'all'
          })}`}
        >
          See all
        </Link>
      </div>
      <DecorativeDash width={30} />
      <div className={styles.list}>
        {LOADING_STATE === 'isLoading' && <div>Loading</div>}
        {LOADING_STATE === 'isLoaded' && (
          <div>
            {!isEmpty(props.documents) && props.documents.items.length > 0 ? (
              props.documents.items.map((doc, index) => {
                const currentFile = getCurrentFile(doc.files)
                let effectiveDate
                if (currentFile && currentFile.effectiveDate) {
                  effectiveDate = currentFile.effectiveDate
                }

                // Add a prefix to SOPs with the format {DOC_TYPE} {DOC_NUMBER} ({DOC_VERSION}) - {DOC_TITLE}
                let titlePrefix = ''
                if (doc.documentIdType === 'SOP' && !isEmpty(doc.documentIdNumber)) {
                  titlePrefix = doc.documentIdType + ' ' + doc.documentIdNumber + ' - '
                }
                const linkTitle =
                  /* eslint-disable-next-line no-magic-number */
                  titlePrefix +
                  (doc.title.length > MAX_TITLE_LENGTH
                    ? doc.title.slice(0, MAX_TITLE_LENGTH + 10) + '...'
                    : doc.title)
                return (
                  <div key={index}>
                    <Link data-testid="document-link" to={doc.url}>
                      {linkTitle}
                    </Link>
                    {effectiveDate && (
                      <div data-testid="document-date" className={styles.date}>
                        {formatDate(effectiveDate)}
                      </div>
                    )}
                  </div>
                )
              })
            ) : (
              <div data-testid="no-results">
                <p>No documents found</p>
              </div>
            )}
          </div>
        )}
      </div>
    </div>
  )
}

const RatesCard = props => {
  return (
    <div className={props.className}>
      <h4 className={styles.title}>Rates</h4>
      <DecorativeDash width={30} />
      <div className={styles.list}>
        {props.rate.map((rate, index) => {
          return (
            <div key={index} className={styles.rateContainer}>
              {rate.name}
              <div className={styles.rate}>{rate.percent}%</div>
            </div>
          )
        })}
      </div>
    </div>
  )
}

const ArticlesCard = props => {
  const eventCategory = `${props.sectionHeaderText.toLowerCase()}-module`
  const { articles } = props

  return (
    <div data-testid="articles-card" className={props.className}>
      <div className={styles.titleContainer}>
        <h4 data-testid="header" className={styles.title}>
          {props.sectionHeaderText}
        </h4>
        <Link
          data-testid="see-all-link"
          to={`/article?${queryString.stringify({
            articleCategory: props.articleCategory,
            program: props.articleProgram
          })}`}
          className={styles.seeAll}
        >
          See all
        </Link>
      </div>
      <DecorativeDash width={30} />
      <div className={styles.list}>
        {articles && articles.items.length ? (
          <div>
            {articles.items.map((article, index) => {
              const linkTitle =
                /* eslint-disable-next-line no-magic-number */
                article.title.length > MAX_TITLE_LENGTH
                  ? article.title.slice(0, MAX_TITLE_LENGTH + 10) + '...'
                  : article.title

              return (
                <div key={index}>
                  <Link data-testid="article-url" to={article.url}>
                    {linkTitle}
                  </Link>
                  <div data-testid="article-date" className={styles.date}>
                    {moment.unix(article.updated).format('MMM D, YYYY')}
                  </div>
                </div>
              )
            })}
          </div>
        ) : (
          <div>loading</div>
        )}
      </div>
    </div>
  )
}

QuickLinks.defaultProps = {
  data: {
    type: 'quickLinks',
    typeOfLinks: [
      {
        type: 'documentLookup',
        documentActivity: [],
        documentProgram: ['CDC/504'],
        documentType: ['SOP'],
        sectionHeaderText: 'SOPs'
      },
      {
        type: 'documentLookup',
        documentActivity: [],
        documentProgram: ['7(a)'],
        documentType: ['Information notice'],
        sectionHeaderText: 'Policy guidance'
      },
      {
        type: 'ratesList',
        rate: [
          {
            type: 'rate',
            name: 'SBA LIBOR Base Rate',
            percent: 4.08
          },
          {
            type: 'rate',
            name: 'SBA Peg Rate',
            percent: 6.08
          },
          {
            type: 'rate',
            name: 'SBA FIXED Base Rate',
            percent: 2.625
          }
        ],
        sectionHeaderText: 'Rates'
      },
      {
        type: 'articleLookup',
        articleProgram: null,
        sectionHeaderText: 'Articles'
      }
    ]
  }
}

QuickLinks.propTypes = {
  data: PropTypes.shape({
    typeOfLinks: PropTypes.arrayOf(
      PropTypes.shape({
        articleCategory: PropTypes.arrayOf(PropTypes.string),
        articleProgram: PropTypes.arrayOf(PropTypes.string),
        documentActivity: PropTypes.arrayOf(PropTypes.string),
        documentProgram: PropTypes.arrayOf(PropTypes.string),
        documentType: PropTypes.arrayOf(PropTypes.string),
        sectionHeaderText: PropTypes.string,
        type: PropTypes.oneOf(['articleLookup', 'documentLookup', 'ratesList'])
      })
    )
  })
}

export { QuickLinks }

export default QuickLinks
