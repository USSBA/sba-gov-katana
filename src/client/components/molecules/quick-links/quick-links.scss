.collection {
  @include clearfix;

  margin-bottom: 80px;

  hr {
    margin: 15px 0;
  }
}

.card {
  border-top: 3px solid $sba-blue;
  margin-bottom: 1.5em;
  margin-top: 1.5em;
  padding-top: 20px;

  @include for-small-and-medium-only {
    @include grid-column(6);
  }
}

.oneCard {
  @include for-large-up {
    @include grid-column(12);
  }
}
.twoCards {
  @include for-large-up {
    @include grid-column(6);
  }
}
.manyCards {
  @include for-large-up {
    @include grid-column(4);
  }
}

.titleContainer {
  position: relative;

  a {
    bottom: 1px;
    right: 0;
    position: absolute;
  }
}

.title {
  display: inline-block;
  margin-bottom: 0px;
}

.date {
  font-style: italic;
  margin-bottom: 15px;
}

.rateContainer {
  margin-bottom: 20px;
}

.rate {
  display: inline-block;
  float: right;
}
