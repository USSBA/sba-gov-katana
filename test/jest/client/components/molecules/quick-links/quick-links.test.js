import React from 'react'
import { QuickLinks } from 'client/components/molecules/quick-links/quick-links.jsx'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

jest.mock('client/services/client-config.js', function() {
  return {
    googleAnalytics: {
      enabled: false
    }
  }
})

const fetchContentIfNeeded = jest.fn()

describe('QuickLinks', () => {
  test('to match snapshot', () => {
    const testProps = {
      actions: {
        fetchContentIfNeeded: fetchContentIfNeeded
      },
      navigation: {
        locationChange: {}
      }
    }

    const mockItem = {
      type: 'quickLinks',
      typeOfLinks: [
        {
          type: 'documentLookup',
          documentActivity: [],
          documentProgram: ['CDC/504'],
          documentType: ['SOP'],
          sectionHeaderText: 'SOPs'
        },
        {
          type: 'documentLookup',
          documentActivity: [],
          documentProgram: ['7(a)'],
          documentType: ['Information notice'],
          sectionHeaderText: 'Policy guidance'
        },
        {
          type: 'ratesList',
          rate: [
            {
              type: 'rate',
              name: 'SBA LIBOR Base Rate',
              percent: 4.08
            },
            {
              type: 'rate',
              name: 'SBA Peg Rate',
              percent: 6.08
            },
            {
              type: 'rate',
              name: 'SBA FIXED Base Rate',
              percent: 2.625
            }
          ],
          sectionHeaderText: 'Rates'
        },
        {
          type: 'articleLookup',
          articleProgram: null,
          sectionHeaderText: 'Articles'
        }
      ]
    }

    const component = renderer.create(
      <QuickLinks data={mockItem} {...testProps} />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
