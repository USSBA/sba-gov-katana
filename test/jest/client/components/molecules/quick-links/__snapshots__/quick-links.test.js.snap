// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickLinks snapshot tests Future effective date is not displayed 1`] = `
<div
  className="collection"
  data-testid="quick-links"
>
  <div
    className="card oneCard"
    data-testid="documents-card"
  >
    <div
      className="titleContainer"
    >
      <h4
        className="title"
        data-testid="header"
      >
        SOPs
      </h4>
      <a
        data-testid="see-all-link"
        onClick={[Function]}
        style={Object {}}
        target="_self"
      >
        See all
      </a>
    </div>
    <hr
      className="dash dash"
      style={
        Object {
          "width": "30px",
        }
      }
    />
    <div>
      <div>
        loading
      </div>
    </div>
  </div>
</div>
`;

exports[`QuickLinks snapshot tests Most recent effective date displayed 1`] = `
<div
  className="collection"
  data-testid="quick-links"
>
  <div
    className="card oneCard"
    data-testid="documents-card"
  >
    <div
      className="titleContainer"
    >
      <h4
        className="title"
        data-testid="header"
      >
        SOPs
      </h4>
      <a
        data-testid="see-all-link"
        onClick={[Function]}
        style={Object {}}
        target="_self"
      >
        See all
      </a>
    </div>
    <hr
      className="dash dash"
      style={
        Object {
          "width": "30px",
        }
      }
    />
    <div>
      <div>
        loading
      </div>
    </div>
  </div>
</div>
`;

exports[`QuickLinks snapshot tests No SOP number so it is not displayed 1`] = `
<div
  className="collection"
  data-testid="quick-links"
>
  <div
    className="card oneCard"
    data-testid="documents-card"
  >
    <div
      className="titleContainer"
    >
      <h4
        className="title"
        data-testid="header"
      >
        SOPs
      </h4>
      <a
        data-testid="see-all-link"
        onClick={[Function]}
        style={Object {}}
        target="_self"
      >
        See all
      </a>
    </div>
    <hr
      className="dash dash"
      style={
        Object {
          "width": "30px",
        }
      }
    />
    <div>
      <div>
        loading
      </div>
    </div>
  </div>
</div>
`;

exports[`QuickLinks snapshot tests SOP number exists and is displayed 1`] = `
<div
  className="collection"
  data-testid="quick-links"
>
  <div
    className="card oneCard"
    data-testid="documents-card"
  >
    <div
      className="titleContainer"
    >
      <h4
        className="title"
        data-testid="header"
      >
        SOPs
      </h4>
      <a
        data-testid="see-all-link"
        onClick={[Function]}
        style={Object {}}
        target="_self"
      >
        See all
      </a>
    </div>
    <hr
      className="dash dash"
      style={
        Object {
          "width": "30px",
        }
      }
    />
    <div>
      <div>
        loading
      </div>
    </div>
  </div>
</div>
`;
